osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Online.Spectator;
+using osu.Game.Screens.OnlinePlay.Multiplayer.Spectate.Sync;
+using osu.Game.Screens.Play;
+using osu.Game.Screens.Spectate;
+
+namespace osu.Game.Screens.OnlinePlay.Multiplayer.Spectate
+{
+    /// <summary>
+    /// A <see cref="SpectatorScreen"/> that spectates multiple users in a match.
+    /// </summary>
+    public class MultiSpectatorScreen : SpectatorScreen
+    {
+        // Isolates beatmap/ruleset to this screen.
+        public override bool DisallowExternalBeatmapRulesetChanges => true;
+
+        /// <summary>
+        /// Whether all spectating players have finished loading.
+        /// </summary>
+        public bool AllPlayersLoaded => instances.All(p => p?.PlayerLoaded == true);
+
+        [Resolved]
+        private SpectatorStreamingClient spectatorClient { get; set; }
+
+        private readonly PlayerArea[] instances;
+        private MasterGameplayClockContainer masterClockContainer;
+        private ISyncManager syncManager;
+        private PlayerGrid grid;
+        private MultiSpectatorLeaderboard leaderboard;
+        private PlayerArea currentAudioSource;
+
+        /// <summary>
+        /// Creates a new <see cref="MultiSpectatorScreen"/>.
+        /// </summary>
+        /// <param name="userIds">The players to spectate.</param>
+        public MultiSpectatorScreen(int[] userIds)
+            : base(userIds.AsSpan().Slice(0, Math.Min(16, userIds.Length)).ToArray())
+        {
+            instances = new PlayerArea[UserIds.Length];
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Container leaderboardContainer;
+            masterClockContainer = new MasterGameplayClockContainer(Beatmap.Value, 0);
+
+            InternalChildren = new[]
+            {
+                (Drawable)(syncManager = new CatchUpSyncManager(masterClockContainer)),
```

Comment:
Bit weird this, haven't seen that done much. Usually `Drawable` is hoisted out as the type of the array (so `new Drawable[] { ... }`)

Reply:
As it is, it would give an R# inspection if I were to specify the array type. Would have to not use the interface and not cast to `Drawable` for that to work :(

Reply:
Oh it's the compound assignment that does it... My bad, that's what I get for only reading the diff online without an IDE.

Reply:
Shall we just change the field to the actual type rather than interface until we need it to be something else?